### ISSUE TITLE
targets-last-text-object does not work for Visual state

### ISSUE THREAD

**Issue by Townk** [noctuid/targets.el#21]
If I select a word with "viw" and later, try to select another word with "v<RET>", I get this error:

```emacs-lisp
Debugger entered--Lisp error: (wrong-type-argument commandp nil)
  call-interactively(nil)
  targets-last-text-object()
  funcall-interactively(targets-last-text-object)
  call-interactively(targets-last-text-object nil nil)
  command-execute(targets-last-text-object)
```

Another related issue I found, is if I try to call `targets-last-text-object` before I use any operator, it also fails in the same way. I believe the `targets-last-text-object` function should simply issue a message when no operator was used before.

**Comment by noctuid** [comment]
`targets-last-text-object` only works for text objects defined with targets.el, which is what I assume causes this problem (let me know if this is not the case). Targets is meant to replace all the default evil text objects (eventually at least).

The other case where there is no last or default object should be fixed.

**Comment by Townk** [comment]
I don't think this is solved. Here is the configuration I have:

```emacs-lisp
    (targets-define-to org-italic "/" nil quote :hooks (org-mode-hook))
    (targets-define-to org-bold "*" nil quote :hooks (org-mode-hook))
    (targets-define-to org-emphasis "_" nil quote :hooks (org-mode-hook))
    (targets-define-to org-description ":" nil quote :hooks (org-mode-hook))

    (general-define-key :keymaps 'global
                        :states '(visual operator)
                        "RET" #'targets-last-text-object)
```

When I'm on an org file, I have the following text:

```
foo foo foo foo foo foo foo /bar |bar bar/ foo foo foo foo foo foo
foo /foobar foobar/ foo foo foo foo foo foo foo foo foo foo
```
(The `|` character is my cursor)

I press: `viw` and I get this:

```
foo foo foo foo foo foo foo /[bar bar bar|]/ foo foo foo foo foo foo
foo /foobar foobar/ foo foo foo foo foo foo foo foo foo foo
```
(The `[  ]` is the visual selection)

Now I press <ESC> and move the cursor to the `foobar` word:

```
foo foo foo foo foo foo foo /bar bar bar/ foo foo foo foo foo foo
foo /foobar| foobar/ foo foo foo foo foo foo foo foo foo foo
```

If I press: `v<RET>`

Only the character under the cursor is selected:

```
foo foo foo foo foo foo foo /bar bar bar/ foo foo foo foo foo foo
foo /foobar[ ]|foobar/ foo foo foo foo foo foo foo foo foo foo
```

Now, this is interesting. If I use any other target with the delete command, the `targets-last-text-object` works as expected. For instance:

```emacs-lisp
    (targets-define-to| org-italic "/" nil quote :hooks (org-mode-hook))
    (targets-define-to org-bold "*" nil quote :hooks (org-mode-hook))
    (targets-define-to org-emphasis "_" nil quote :hooks (org-mode-hook))
    (targets-define-to org-description ":" nil quote :hooks (org-mode-hook))
```

When I press `di(`, I got this:

```emacs-lisp
    (|)
    (targets-define-to org-bold "*" nil quote :hooks (org-mode-hook))
    (targets-define-to org-emphasis "_" nil quote :hooks (org-mode-hook))
    (targets-define-to org-description ":" nil quote :hooks (org-mode-hook))
```

Let's say I move the cursor couple lines below:

```emacs-lisp
    ()
    (targets-define-to org-bold "*" nil quote :hooks (org-mode-hook))
    (|targets-define-to org-emphasis "_" nil quote :hooks (org-mode-hook))
    (targets-define-to org-description ":" nil quote :hooks (org-mode-hook))
```

And if I press `d<RET>`, the behavior is what I would expect:

```emacs-lisp
    ()
    (targets-define-to org-bold "*" nil quote :hooks (org-mode-hook))
    (|)
    (targets-define-to org-description ":" nil quote :hooks (org-mode-hook))
```

Am I misunderstanding something here?

**Comment by Townk** [comment]
BTW, I was looking at the code and notice that you make a distinction between Visual operators and the rest. Why there is this difference? Wouldn't be better to let operation share the movement? I remove that distinction and now everything works as expected including the visual operator!

I created this PR with my changes in case you're interested: https://github.com/noctuid/targets.el/pull/22

**Comment by noctuid** [comment]
The last visual text object is cleared when exiting visual state. The idea is that there often isn't connection between the last text object used with an operator and what text object you want to use next in visual state. Using the "last" text object in visual state is more geared towards a single session where you incrementally expand the region. If the last text object used with an operator is already correct for the next operation, I wouldn't use visual state. If I wanted to do something more complicated, I may not even remember the last text object, and to me, it makes more sense to initially rely on a well-chosen default text object or alternatively run a specific one once in visual state and continued to expand afterwards with the shorter keybinding.

If you would prefer the history to be shared, I can add an option for it.

**Comment by Townk** [comment]
The reason why I want to share the text objects is to do destructive operations in 2 steps. First I would use the Visual operator to check if my destructive operation will affect the correct area. Then I would simply do a `d<RET>` to remove it or something else.

The core idea here is the same as the `substitute` command (at least that is how it's called on Vim). You can do a **search** first, then you can use empty match with `substitute` to apply the changes (e.g. `%s//new text/g`).

Also, this is the PR that only adds an option to targets: https://github.com/noctuid/targets.el/pull/23

BTW, when are we going to see targets on MELPA?

**Comment by noctuid** [comment]
If you had the correct area selected in visual mode, wouldn't you just press `d`? Or are you saying that it's useful because you may want to repeatedly use the same text object afterwards? I still think it's a good idea to add the setting for sharing the history, but it may be worth noting that visual repeating by action (not changed text) is possible in evil by advising/overriding a few functions and will hopefully be added as an option at some point (see [here](https://github.com/emacs-evil/evil/issues/742)).

I don't have a lot of time to work on my packages recently and am prioritizing some others already on MELPA, so it will probably be a while. I want to at least fix the obvious bugs/issues mentioned in #16 first.
